packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts

```diff
@@ -394,6 +403,31 @@ export class UserPoolClient extends Resource implements IUserPoolClient {
 
     this.userPoolClientId = resource.ref;
     this._userPoolClientName = props.userPoolClientName;
+
+    if (props.generateSecret) {
+      const describeCognitoUserPoolClient = new AwsCustomResource(
+        this,
+        'DescribeCognitoUserPoolClient',
+        {
+          resourceType: 'Custom::DescribeCognitoUserPoolClient',
+          onCreate: {
+            region: Stack.of(this).region,
+            service: 'CognitoIdentityServiceProvider',
+            action: 'describeUserPoolClient',
+            parameters: {
+              UserPoolId: props.userPool.userPoolId,
+              ClientId: this.userPoolClientId,
+            },
+            physicalResourceId: PhysicalResourceId.of(this.userPoolClientId),
+          },
+          policy: AwsCustomResourcePolicy.fromSdkCalls({
+            resources: AwsCustomResourcePolicy.ANY_RESOURCE,
+          }),
+        },
+      );
+
+      this.userPoolClientSecret = describeCognitoUserPoolClient.getResponseField('UserPoolClient.ClientSecret');
+    }
```

Comment:
I like the idea of supporting this at the L2 level via `userPoolClientSecret`, however, I feel strongly that we ought to produce the custom resource if and only if the `userPoolClientSecret` is accessed. This would imply creating this custom resource in a getter function (`get userPoolClientSecret(): string { ... }`).

Reply:
Good suggestion. Refactored to only create the Custom Resource if the property is accessed via a getter.